cns/restserver/ipam.go

```diff
@@ -12,76 +12,97 @@ import (
+	if returnCode != types.Success {
+		reserveResp := &cns.IPConfigResponse{
+			Response: cns.Response{
+				ReturnCode: returnCode,
+				Message:    returnMessage,
+			},
 		}
+		err = service.Listener.Encode(w, &reserveResp)
+		logger.ResponseEx(service.Name+operationName, ipconfigRequest, reserveResp, reserveResp.Response.ReturnCode, err)
+		return
 	}
 
-	resp := cns.Response{
-		ReturnCode: returnCode,
-		Message:    returnMessage,
+	// record a pod requesting an IP
+	service.podsPendingIPAllocation.Push(podInfo.Key())
+
+	podIPInfo, err := requestIPConfigHelper(service, ipconfigRequest)
+	if err != nil {
+		reserveResp := &cns.IPConfigResponse{
+			Response: cns.Response{
+				ReturnCode: types.FailedToAllocateIPConfig,
+				Message:    fmt.Sprintf("AllocateIPConfig failed: %v, IP config request is %s", err, ipconfigRequest),
+			},
+			PodIpInfo: podIPInfo,
+		}
+		err = service.Listener.Encode(w, &reserveResp)
+		logger.ResponseEx(service.Name+operationName, ipconfigRequest, reserveResp, reserveResp.Response.ReturnCode, err)
+		return
 	}
 
+	// record a pod allocated an IP
+	defer func() {
+		// observe allocation wait time
+		ipAllocationLatency.Observe(float64(service.podsPendingIPAllocation.Pop(podInfo.Key())))
+	}()
 	reserveResp := &cns.IPConfigResponse{
-		Response: resp,
+		Response: cns.Response{
+			ReturnCode: types.Success,
+		},
+		PodIpInfo: podIPInfo,
 	}
-	reserveResp.PodIpInfo = podIpInfo
 
 	err = service.Listener.Encode(w, &reserveResp)
-	logger.ResponseEx(service.Name+operationName, ipconfigRequest, reserveResp, resp.ReturnCode, err)
+	logger.ResponseEx(service.Name+operationName, ipconfigRequest, reserveResp, reserveResp.Response.ReturnCode, err)
 }
 
 func (service *HTTPRestService) releaseIPConfigHandler(w http.ResponseWriter, r *http.Request) {
-	req := cns.IPConfigRequest{}
-	resp := cns.Response{}
-	var err error
-
-	defer func() {
-		err = service.Listener.Encode(w, &resp)
-		logger.ResponseEx(service.Name, req, resp, resp.ReturnCode, err)
-	}()
-
-	err = service.Listener.Decode(w, r, &req)
+	var req cns.IPConfigRequest
+	err := service.Listener.Decode(w, r, &req)
 	logger.Request(service.Name+"releaseIPConfigHandler", req, err)
 	if err != nil {
-		resp.ReturnCode = types.UnexpectedError
-		resp.Message = err.Error()
+		resp := cns.Response{
+			ReturnCode: types.UnexpectedError,
+			Message:    err.Error(),
+		}
 		logger.Errorf("releaseIPConfigHandler decode failed becase %v, release IP config info %s", resp.Message, req)
+		err = service.Listener.Encode(w, &resp)
+		logger.ResponseEx(service.Name, req, resp, resp.ReturnCode, err)
 		return
 	}
 
-	var podInfo cns.PodInfo
-	podInfo, resp.ReturnCode, resp.Message = service.validateIPConfigRequest(req)
+	podInfo, returnCode, message := service.validateIPConfigRequest(req)
+
+	// remove this pod from the waiting for IP allocation set, if it is present
+	_ = service.podsPendingIPAllocation.Pop(podInfo.Key())
```

Comment:
Actually with data deleting during release wont give us complete data. AFAIK, ContainerD calls ADD and if ADD fails to provision then it calls DEL followed by another ADD. If that is true, then we will never be able to truely evaluate the actual time it took to allocate an IP

Reply:
I have updated the PR to use a custom queue-style structure for this:
- We will hold up to the last 250 (maxpods) unique pod IDs that we have seen, recording the timestamp that that pod was first seen.
- When an IP is allocated, we will pop the entry for that pod and record the duration since it was first seen.
- When the queue is full and we see a new pod, we will drop the oldest pod record by first-seen-timestamp and record the new pod.